File: hedera-mirror-rest/model/transaction.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {filterKeys} from '../constants';

class Transaction {
  static tableAlias = 't';
  static tableName = 'transaction';
  static CHARGED_TX_FEE = `charged_tx_fee`;
  static CONSENSUS_TIMESTAMP = `consensus_timestamp`;
  static ENTITY_ID = `entity_id`;
  static INITIAL_BALANCE = `initial_balance`;
  static MAX_FEE = `max_fee`;
  static MEMO = `memo`;
  static NODE_ACCOUNT_ID = `node_account_id`;
  static NONCE = `nonce`;
  static PARENT_CONSENSUS_TIMESTAMP = `parent_consensus_timestamp`;
  static PAYER_ACCOUNT_ID = `payer_account_id`;
  static RESULT = `result`;
  static SCHEDULED = `scheduled`;
  static TRANSACTION_HASH = `transaction_hash`;
  static TRANSACTION_BYTES = `transaction_bytes`;
  static TYPE = `type`;
  static VALID_DURATION_SECONDS = `valid_duration_seconds`;
  static VALID_START_NS = `valid_start_ns`;
  static INDEX = `index`;
  static FILTER_MAP = {
    [filterKeys.TIMESTAMP]: Transaction.getFullName(Transaction.CONSENSUS_TIMESTAMP),
  };

  /**
   * Parses transaction table columns into object
   */
  constructor(transaction) {
    this.chargedTxFee = transaction.charged_tx_fee;
    this.consensusTimestamp = transaction.consensus_timestamp;
    this.entityId = transaction.entity_id;
    this.initialBalance = transaction.initial_balance;
    this.maxFee = transaction.max_fee;
    this.memo = transaction.memo;
    this.nodeAccountId = transaction.node_account_id;
    this.nonce = transaction.nonce;
    this.parentConsensusTimestamp = transaction.parent_consensus_timestamp;
    this.payerAccountId = transaction.payer_account_id;
    this.result = transaction.result;
    this.scheduled = transaction.scheduled;
    this.transactionHash = transaction.transaction_hash;
    this.transactionBytes = transaction.transaction_bytes;
    this.type = transaction.type;
    this.validDurationSeconds = transaction.valid_duration_seconds;
    this.validStartNs = transaction.valid_start_ns;
    this.index = transaction.index;
  }

  /**
   * Gets full column name with table alias prepended.
   *
   * @param {string} columnName
   * @private
   */
  static getFullName(columnName) {
    return `${this.tableAlias}.${columnName}`;
  }
}

export default Transaction;

File: hedera-mirror-rosetta/app/persistence/domain/transaction.go
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

package domain

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"fmt"
)

const (
	TransactionTypeCryptoCreateAccount int16 = 11
	TransactionTypeCryptoTransfer      int16 = 14
	TransactionTypeTokenCreation       int16 = 29
	TransactionTypeTokenDeletion       int16 = 35
	TransactionTypeTokenUpdate         int16 = 36
	TransactionTypeTokenMint           int16 = 37
	TransactionTypeTokenDissociate     int16 = 41

	transactionTableName = "transaction"
)

type ItemizedTransfer struct {
	Amount     int64    `json:"amount"`
	EntityId   EntityId `json:"entity_id"`
	IsApproval bool     `json:"is_approval"`
}

type ItemizedTransferSlice []ItemizedTransfer

func (i *ItemizedTransferSlice) Scan(value interface{}) error {
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New(fmt.Sprint("Failed to unmarshal JSONB value", value))
	}

	result := ItemizedTransferSlice{}
	err := json.Unmarshal(bytes, &result)
	*i = result
	return err
}

func (i ItemizedTransferSlice) Value() (driver.Value, error) {
	if len(i) == 0 {
		return nil, nil
	}

	return json.Marshal(i)
}

type JSONB map[string]interface{}

type Transaction struct {
	ConsensusTimestamp       int64 `gorm:"primaryKey"`
	ChargedTxFee             int64
	EntityId                 *EntityId
	Errata                   *string
	InitialBalance           int64
	ItemizedTransfer         ItemizedTransferSlice `gorm:"type:jsonb"`
	MaxFee                   int64
	Memo                     []byte
	NodeAccountId            *EntityId
	Nonce                    int32
	ParentConsensusTimestamp int64
	PayerAccountId           EntityId
	Result                   int16
	Scheduled                bool
	TransactionBytes         []byte
	TransactionHash          []byte
	Type                     int16
	ValidDurationSeconds     int64
	ValidStartNs             int64
}

func (Transaction) TableName() string {
	return transactionTableName
}

func (itemizedTransfer JSONB) Value() (driver.Value, error) {
	return json.Marshal(itemizedTransfer)
}

func (itemizedTransfer *JSONB) Scan(value interface{}) error {
	data, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}
	return json.Unmarshal(data, &itemizedTransfer)
}
